### PR TITLE
[#1193] Adjust runLenghtEncoding to match schema in #735

### LINKED ISSUE #1193: Verify run length encoding in Wallet/KeyUtils with BOLOS app and BIP44 implementation
We need to ensure that 

1) Our run length encoding is to the exact specifications here: https://github.com/Constellation-Labs/constellation/issues/735#issuecomment-619652684
2) Our BIP44 implementation https://github.com/Constellation-Labs/constellation/pull/1086 is compatible with the BOLOS app

### PR DESCRIPTION
**Pull request by TheMMaciek**
Work in progress. Of course changes need to be also ported to main project. Failing test because the test transaction was generated with old runLengthEncoding logic. After porting changes to main project test can be fixed.

Resolves #1193

### CODE REVIEW

File: wallet/src/main/scala/org/constellation/wallet/Utils.scala [hunk 1]
```diff
@@ -100,6 +102,32 @@ object KryoSerializer {
     kryoPool.fromBytes(bytes).asInstanceOf[T]
 }
 
+sealed trait EncodableValue// {val value: Any}
```
Comment (marcinwadon) [inline wallet/src/main/scala/org/constellation/wallet/Utils.scala:105]: `extends AnyVal`?

Comment (TheMMaciek) [reply wallet/src/main/scala/org/constellation/wallet/Utils.scala:105]: It doesn't pass with extends AnyVal - illegal inheritance from a value class. The idea here was to enforce creating subclasses with a value field of arbitrary type. But `Any` looks not that pretty so it's commented out - but I think I will uncomment it to have some suggestion on what field we want.

Comment (marcinwadon) [reply wallet/src/main/scala/org/constellation/wallet/Utils.scala]: so maybe a generic value will work?

File: src/test/scala/org/constellation/domain/checkpointBlock/CheckpointBlockDoubleSpendCheckerTest.scala [hunk 2]
```diff
@@ -25,7 +25,7 @@ class CheckpointBlockDoubleSpendCheckerTest extends FunSuite with BeforeAndAfter
   }
 
   test("it should return transaction when transaction's ordinal number is less than or equal the last accepted one") {
-    val firstTransaction = createTransaction("src", "firstTransaction", 1)
+    val firstTransaction = createTransaction("src", "abc123", 1)
```
Comment (buckysballs) [inline src/test/scala/org/constellation/domain/checkpointBlock/CheckpointBlockDoubleSpendCheckerTest.scala:28]: Just a heads up that we can rely on the default value of "abc123" here

File: src/test/scala/org/constellation/wallet/ValidateWalletFuncTest.scala [hunk 3]
```diff
@@ -39,6 +40,92 @@ class ValidateWalletFuncTest extends FlatSpec {
...
+      Seq(
+        TypedEdgeHash(src, EdgeHashType.AddressHash),
+        TypedEdgeHash(dst, EdgeHashType.AddressHash)
+      ),
+      TypedEdgeHash(txData.getEncoding, EdgeHashType.TransactionDataHash)
+    )
+    val soe = SignHelp.signedObservationEdge(oe)(Fixtures.tempKey)
+    val tx = Transaction(edge = Edge(oe, soe, txData), lastTxRef = lastTxRef, isDummy = false, isTest = false)
+    val expectedEncoded =
+      "02" + //number of parents
```
Comment (buckysballs) [inline src/test/scala/org/constellation/wallet/ValidateWalletFuncTest.scala:63]: If we're re-using these below consider making each part of expectedEncoded vals

File: wallet/src/main/scala/org/constellation/wallet/Utils.scala [hunk 4]
```diff
@@ -100,6 +102,32 @@ object KryoSerializer {
...
+    case EncodableBase16(value) =>
+      val toEncode = if(value.isEmpty) Array.emptyByteArray else BigInt(value, 16).toByteArray
+      encodeLength(toEncode.length) + value
+    case EncodableASCII(value) =>
+      val toEncode = value.getBytes(US_ASCII)
+      encodeLength(toEncode.length) + bytes2hex(toEncode)
+  }
+
+  def runLengthEncoding(values: Seq[EncodableValue]): String = values.foldLeft("") {
+    (acc, value) => acc + encode(value)
```
Comment (buckysballs) [inline wallet/src/main/scala/org/constellation/wallet/Utils.scala:123]: Nice

File: wallet/src/main/scala/org/constellation/wallet/Utils.scala (hunk 1 above)
Comment (buckysballs) [reply wallet/src/main/scala/org/constellation/wallet/Utils.scala]: Yeah we'll want a generic for serialization eventually

File: src/test/scala/org/constellation/domain/checkpointBlock/CheckpointBlockDoubleSpendCheckerTest.scala (hunk 2 above)
Comment (TheMMaciek) [reply src/test/scala/org/constellation/domain/checkpointBlock/CheckpointBlockDoubleSpendCheckerTest.scala:28]: Do you mean that we can expect it to be base16? If yes I would add simple static validation when submitting a transaction ensuring that prevHash can be parsed as base16.

File: src/test/scala/org/constellation/wallet/ValidateWalletFuncTest.scala (hunk 3 above)
Comment (TheMMaciek) [reply src/test/scala/org/constellation/wallet/ValidateWalletFuncTest.scala:63]: I'm reusing some of it, but I think actually that extracting common parts in this particular example can make things less clear and I would need to change from flatSpec to freeSpec hence refactor other tests to extract these - I think it's not worth it ;).

File: src/main/scala/org/constellation/primitives/Schema.scala [hunk 5]
```diff
@@ -249,9 +250,9 @@ object Schema {
     fee: Option[Long] = None,
     salt: Long = Random.nextLong()
   ) extends Signable {
-    override def getEncoding = {
-      val encodedAmount = runLengthEncoding(Seq(amount.toHexString): _*)
-      val encodedFeeSalt = runLengthEncoding(Seq(fee.getOrElse(0L).toString, salt.toHexString): _*)
+    override def getEncoding: String = {
```
Comment (buckysballs) [inline src/main/scala/org/constellation/primitives/Schema.scala:253]: @ffox77 ^

**Pull request closed without merging**